src/lib/plugins/GraphQLPlugin/client/GraphQLQueryService.ts

```diff
@@ -0,0 +1,124 @@
+import { DocumentNode, GraphQLError } from 'graphql'
+import { ApolloClient, WatchQueryOptions, ApolloError } from 'apollo-client'
+import { Subscription } from 'apollo-client/util/Observable'
+import { isUndefined } from 'util'
+import { Dispatch } from 'redux'
+import { Service, decorateServiceProperty } from '../../../core/Service'
+import { injectDependency } from '../../../core/InjectionClient'
+import { createSelectService, StateSelector } from '../../StorePlugin/SelectService'
+import { injectDispatch } from '../../StorePlugin/StorePlugin'
+import GraphQLPlugin, { GraphQLAction } from '../GraphQLPlugin.common'
+
+export interface GraphQLQueryService<Result> extends Service {
+  readonly data: Result
+  readonly optionalData: Result | undefined
+}
+
+export interface GraphQLQueryServiceState {
+  data?: any
+  stale?: boolean
+  loading?: boolean
+  errors?: GraphQLError[]
+}
+
+export function decorateGraphQLQuery(query: DocumentNode) {
+  class GraphQLQueryServiceImpl<Result> extends Service<GraphQLQueryServiceState> {
+    private queryObserver: Subscription
+
+    @injectDependency(ApolloClient)
+    client: ApolloClient<any>
+
+    @createSelectService((state, props) => GraphQLPlugin.selectErrors({ variables: props, query })(state))
+    errors: StateSelector<GraphQLError[]>
+
+    @injectDispatch
+    dispatch: Dispatch<GraphQLAction>
+
+    get data(): Result {
+      const data = this.optionalData
+      if (!data) {
+        throw new Error([
+          `GraphQL data for this component is not yet available.`,
+          `You should avoid calling query.data in componentWillMount, serviceWillLoad`,
+          `or a constructor`
+        ].join(' '))
+      }
+
+      return data
+    }
+
+    get optionalData(): Result | undefined {
+      const errors = this.errors.value
+      if (errors && errors.length > 0) {
+        throw new ApolloError({ graphQLErrors: errors })
+      }
+
+      if (!isUndefined(this.state.data)) {
+        return this.state.data
+      }
+
+      try {
+        return this.client.readQuery(this.queryConfig)
```

Comment:
So, does this means the query will be re-run everytime the data object is used?

Obviously 'most' queries will be cached, but if a query was set not to cache ever, would each use of data (e.g. twice within a function, or on a re-render) retrigger it running?

Reply:
`readQuery()` just looks in the cache. It returns the cached value, if present, or else throws.